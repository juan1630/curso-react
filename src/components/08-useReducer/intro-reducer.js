

const initialState = [{
    id:1,
    todo:"Comprar pan",
    done: false
}];




// pirmer reducer, porque regresa un nuevo estado
const todoReducer = (  state = initialState, action  ) => {
    

    if( action?.type == 'agregar'  ){

        return  [ ...state, action.payload]
    }
    
    return state;

}

// no usar push cuando se tarabaja en react

let todos = todoReducer();


const newTodo = {
    id:2,
    todo:'Comprar leche',
    done: false
}


const action = {
    type:'agregar',
    payload: newTodo
}


todos = todoReducer( todos,  action );

console.log( todos );