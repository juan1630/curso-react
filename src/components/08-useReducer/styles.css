

body {
    padding: 1.1rem;
}


h1 {
    font-size: 1.8rem;
    text-align: center;
    font-weight: 600;
    color: blue;
}


ul{
    font-size: 1.2rem;
    font-weight: 500;
}


.btn-danger {
    margin-left: 1.3rem;
}



.todos-items{
    margin-top: 1.2rem;
}

.agregar{
    text-align: center;

}


.item-todo {
    cursor: pointer;
}


.complete {
    text-decoration: line-through;
}


.form{
    display: block;
    width: 100%;

}

.btn-block{
    width: 100%;
    margin-top: 1.1rem;
}