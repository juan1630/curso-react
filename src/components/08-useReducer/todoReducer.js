


export  const todoReducer = ( state = [] , action ) => {

    switch (action.type) {



        case 'add':
            
            return [...state, action.payload];
        case 'delete':
            return state.filter(  todo => todo.id !== action.payload );  //id = 6684683124

        case 'toggle':
            return state.map( todo => {
                
                if(todo.id === action.payload ){
                    return {
                        ...todo,
                        done: !todo.done
                    }
                }else {
                    return todo;
                }
            } )
        
        default:
            return state;
    }


    // definimos el tipo y si accion de acuerdo a su tipo


}