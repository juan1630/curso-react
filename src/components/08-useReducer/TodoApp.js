import React, { useEffect, useReducer } from 'react';
import {todoReducer } from './todoReducer';

import './styles.css'
import { useForm } from '../../hooks/useHook';



const init = () => {
    // return[{

    //     id:  new Date().getTime(),
    //     desc: 'Aprender React',
    //     done: false
    
    // }];

    return JSON.parse(  localStorage.getItem('todos') ) || [];

}

export const TodoApp = () => {  
    

    const [ todos, dispatch ] =  useReducer(todoReducer, [] , init  );


  const [ { description}, handleInputChange, reset  ] =  useForm({
        
     description:''

    });


    // se usa el useEffect para el localstorage 

    useEffect( () => {
        // cualquier cambio y se vuelve a ejecutar el efecto

        localStorage.setItem('todos', JSON.stringify( todos ));

    }, [ todos ]);


    const handleDelete = (  todoId ) => {
        
        // recibir el id 

        console.log( todoId );

        // crear accion para borrarlo

        const actionDelete = {
            type:'delete',
            payload: todoId
        }

        // dispatch 

        dispatch( actionDelete )
    }


    const handleToggle = (idToggle) => {
        

        const actionToggle = {
            type:'toggle',
            payload: idToggle
        }


        dispatch( actionToggle )
    }

    const handleOnSubmit = (e) => {
        

        if( description.trim().length < 1 ){
            return;
        }

        e.preventDefault();


        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }


        const action ={
            type:'add',
            payload:newTodo
        }

        // el dispatch ya sabe que funcion ejecutar
        dispatch( action );

        // reseteamos el form 
        reset();



    }


    return (
        <div>
            <h1> Todo App (  {todos.length} ) </h1>
            <hr/>

        <div className="row"  >

                <div className="col-7" >

                    <h3>  Todos </h3>

                <ul className="todos-items" >
                {
                    todos.map( (todo, i) => (
                        
                        <li  onClick={ ()=> { handleToggle(todo.id) } } className={ `${todo.done && 'complete'} ` }  key={todo.id} >  { i +1 } - {todo.desc} 
                            <hr/>
                            <button 
                            
                            className="btn btn-danger"
                            onClick={ () => { handleDelete( todo.id ) } }
                            > 

                                Borrar
                            </button>
                        
                          </li>
                    ))
                }
            </ul>


                </div>

                <div className="col-5" >
                   <h3 className="agregar" >
                        Agregar todo
                   </h3>


                   <hr/>

                    <form onSubmit={  handleOnSubmit }  >


                        <input 
                        type="text"
                        className="form-control"
                        name="description"
                        placeholder="Comer..."
                        autoComplete="off"
                        onChange={ handleInputChange }
                        value={ description }
                        />


                        <button className="btn btn-success btn-block" > Agregar </button>

                    </form>

                </div>

        </div>
            
           

        </div>
    )
}
