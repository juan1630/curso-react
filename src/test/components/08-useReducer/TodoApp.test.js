import react from 'react';
import { mount, shallow } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp'
import { demoTodos} from '../../fixture/demoTodos'
import { act } from '@testing-library/react';



describe('pruebas en todoApp', () => {


    const wrapper = shallow(  <TodoApp/> );

    Storage.prototype.setItem = jest.fn( ()=> {} )


    test('Debe de mostrarse correctamente', ()=> {
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe de agregar un Todo ', () => {
        
        const wrapper = mount(<TodoApp/>);
        // se muestra ya rendirizado el componente

        act( ()=> {
            wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] )
            wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[1] )

        } );


        expect( wrapper.find('h1').text().trim() ).toBe('Todo App (      2 )')
        expect( localStorage.setItem ).toHaveBeenCalledTimes(2);

    })


    test('debe de llamar un TODO', () => {
        wrapper.find('TodoAdd').prop('handleAddTodo')( demoTodos[0] );
        expect( wrapper.find('h1').text().trim() ).toBe('Todo App (  1 )')
    })
    
    
})
