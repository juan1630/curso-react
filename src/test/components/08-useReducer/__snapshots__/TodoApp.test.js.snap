// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en todoApp Debe de mostrarse correctamente 1`] = `
<div>
  <h1>
     Todo App (  
    0
     ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <h3>
          Todos 
      </h3>
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
